Refactor surface.py code for easier extension

eepeterson opened the pull request:
This PR is in response to issue #1475 and PR #1474 and should provide a cleaner inheritance structure for Surfaces. The main change is two mixin classes were added for Planes and Quadrics which implement all the common transformations for those types of surfaces and will be extended soon to include rotations. 

There are a number of benefits to this including:
1.) There is no longer an inconsistent internal state for subclasses of various surfaces.
2.) There is only one evaluate method, one translate method (and will be only one rotate method) per surface order i.e (planes = order 1, quadrics = order 2) 
3.) This will make adding additional classes inlcuding elliptical cylinders, torii, etc. easier to implement
4.) It gets the model building closer to MCNP capabilities and will allow for easier translation
5.) It should provide more power during the model building process and make simplifying geometries an easier task.

This is close to being ready to merge. As of right now a few pytest results come out slightly different and I need to implement two more coordinate transformations for general cylinders and general cones. But as of right now neither of these is an acceptable input to the C++ layer anyway so that will need to be changed soon.

As of right now the API is still intact, but I'm suggesting that Surface parameters `boundary_type`, `surface_id` and `name` only be permitted to be passed by keyword argument to the Surface subclasses. This makes it more explicit that anything in **kwargs is being passed up through the mixin classes to Surface and that any positional arguments are implicitly only for the class itself, not its parent. I've implemented a work around and issued a warning in places where this was causing the test suite to fail. I think it improves the readability of the code and greatly reduces the chances of problems arising down the line. I also don't think it should be too disastrous a consequence for the end user, but I'm open to suggestions.

smharper commented:
Thanks for opening this draft PR!  I think you might have the honor of being the first of our team to venture out into the multiple inheritance territory!  This is definitely a good way to express the relationships between the different surface types.

Thinking ahead to C++ changes and how Python should match up with them, I think it would be better to override the `evaluate` function in the concrete classes with their simpler versions.  Using the general quadric evaluate for e.g. z-cylinders would add a lot of flops and might give us a performance regression on the C++ side.  And it's probably best to use the same evaluate functions in C++ and Python so that both implementations will give identical results.

Also, I'd recommend not repeating docstrings when overriding methods like `bounding_box`. Python's `help` function is smart enough to look through an object's base classes to find docstrings, and I assume most other Python tools will do the same.  Removing some of those duplicates will help with the big LoC count.

eepeterson commented:
Thanks for the feedback @smharper! I've gone back through and implemented individual evaluate and translate functions like there were in the beginning. I also removed redundant docstrings. Basically the main thing this PR adds now is the two metaclasses for planes and quadrics and methods for each surface type to be able to translate itself into a more abstract surface which isn't useful for translations, but will be for rotations.

eepeterson commented on openmc/surface.py:
```diff
@@ -936,33 +979,151 @@ class Cylinder(Surface):
+        x0, y0, z0 = p1
+        dx, dy, dz = p2 - p1
+
+        return cls(x0=x0, y0=y0, z0=z0, r=r, dx=dx, dy=dy, dz=dz, **kwargs)
```
I'm planning on implementing the Cylinder and Cone classes in the C++ layer, but perhaps as a second PR? I can override `to_xml_element` for now to raise a `NotImplementedError` or throw it in __init__ for now. I more or less wanted to implement the complete hierarchy in the python API first. Any thoughts?

eepeterson commented:
@paulromano I've flipped flopped back and forth about `translate` especially after @smharper suggested overriding `evaulate` in all concrete class instances for performance so I chose to override `translate` too and went into an existential spiral about the purpose of the mixin haha... , but I think there is an effective way to put it back into the mixin so I'll do that today.

eepeterson commented on openmc/surface.py:
```diff
@@ -268,62 +293,103 @@ def from_hdf5(group):
 
 
-class Plane(Surface):
+class Plane(PlaneMixin, Surface):
```
Since this mixin has the `to_xml_element` code for Planes it needs to come first in the MRO for cooperative inheritance to work, which seems to make sense to me as it is intermediate in the hierarchy between a concrete Plane object and a general surface. It's how I prevent XPlanes etc  from inheriting the attributes of Plane which caused inconsistencies in the first place. The alternative would be to get rid of the mixin, inherit from Plane and override setters to disallow setting properties like a, b,c, d but the mixin seemed like the more elegant solution, but I'd be interested in hearing your thoughts.

eepeterson commented:
I just added `__new__` to `Plane` to return a `(X,Y,Z)Plane` instance if the `a, b, c` values are right. This involved changing the defaults for `Plane` to be `a=b=c=1` because previously the default `Plane` is really an `XPlane` which ended up breaking the `clone` function of `Surface`. Right now all the unit tests pass and playing around with translations and instantiation in ipython all work nicely, but in the `regression_tests/deplete/test.py` I get a seg fault Fatal Python error with almost no information as to why. @paulromano would you have any idea why this is happening? 

The other thing that has been really stumping me for the past few days is that I'm forced to call `__init__` for specific planes inside the `Plane.__new__` method. The docs for `__new__` say: 

"If `__new__()` is invoked during object construction and it returns an instance or subclass of `cls`, then the new instance’s `__init__()` method will be invoked like `__init__(self[, ...])`, where `self` is the new instance and the remaining arguments are the same as were passed to the object constructor." 

I thought that by registering XPlane, YPlane, and ZPlane as virtual subclasses of Plane this would work, but it doesn't. Both `isinstance(obj, Plane)` and `issubclass(obj.__class, Plane)` return `True` even when I call them just before returning the new object in `__new__`, but `Plane.__subclasses__()` is empty. I've been struggling to figure out why this is happening and was wondering if @paulromano or @smharper have any ideas...

eepeterson commented on openmc/surface.py:
```diff
@@ -2274,3 +2509,19 @@ def translate(self, vector, memo=None):
+    return {c._type: c for c in _get_subclasses(cls)}
+
+
+_SURFACE_CLASSES = _get_subclass_map(Surface)
```
I was just covering my bases in case anything in the future inherited from these derived classes rather than surface, but you're right.

paulromano commented:
Taking a look at CPython internals, the way that the subclass relationship is determined after `__new__` is called is by walking up the MRO. So, for `XPlane` to be considered a subclass of `Plane`, `Plane` would have to appear in its MRO, and it currently doesn't:
```Python
>>> openmc.XPlane.__mro__                                                                                                                                                                                          
(<class 'openmc.surface.XPlane'>,
 <class 'openmc.surface.PlaneMixin'>,
 <class 'openmc.surface.Surface'>,
 <class 'openmc.mixin.IDManagerMixin'>,
 <class 'object'>)
```
Registering a class as a virtual subclass does not modify the MRO; this caveat is mentioned explicitly [here](https://docs.python.org/3/library/abc.html#abc.ABCMeta) in the description of ABCMeta.

I'll take a look at the segfault and see if I can figure anything out.

paulromano commented:
Ok, so I figured out what was going on with the depletion problem. It boils down to two issues:
- In `tests/regression_tests/deplete/example_geometry.py`, there's a case where we're calling `Plane(a=...., b=...)`. The new convention of setting `c` to 1 is messing this up because it really should be zero.
- Even with that fixed, some of these planes are getting turned into x-planes/y-planes, but the sense of the surface is reversed. This results in some areas of the problem being undefined, and other areas having multiple cells that overlap.

I'm going to make two recommendations based on this:
1. We shouldn't change the default coefficients of `Plane`. This resulted in a really subtle error, and it's not hard to imagine this causing problems for users who are already (unknowingly or knowingly) relying on the current behavior.
2. We shouldn't auto-convert to the simpler planes because the sense might get messed up. We could restrict it to cases where the surface sense is preserved, but this would create a weird situation where `Plane(a=1, b=0, c=0, d=0)` returns an `XPlane` but `Plane(a=-1, b=0, c=0, d=0)` does not.

eepeterson commented:
@paulromano thanks so much for tracking that down and for all the helpful ideas. I’ll work on fixing everything tomorrow.

drewejohnson commented on tests/regression_tests/deplete/example_geometry.py:
```diff
@@ -188,6 +188,7 @@ def segment_pin(n_rings, n_wedges, r_fuel, r_gap, r_clad):
 
     fuel_wedges = [openmc.Plane(a=math.cos(theta[i]), b=math.sin(theta[i]))
                    for i in range(n_wedges)]
+    print(fuel_wedges)
```
Got a debug statement that should be removed

paulromano commented on openmc/surface.py:
```diff
@@ -936,33 +978,139 @@ class Cylinder(Surface):
+        h = 2*(dy*(z0*dz + x0*dx) - y0*(dx2 + dz2))
+        j = 2*(dz*(y0*dy + x0*dx) - z0*(dx2 + dy2))
+        k = x0*x0*(dy2 + dz2) + y0*y0*(dx2 + dz2) + z0*z0*(dx2 + dy2) \
+            + e*y0*z0 + f*x0*z0 + d*x0*y0 - r*r*(dx2 + dy2 + dz2)
```
Not at all obvious how one gets these coefficients (_I know_, but someone else coming in and looking at this would probably have no idea). I would suggest copying the comment from the `cylinder_from_points` function in `openmc/model/funcs.py` that describes how these are obtained.

Also, as you'll see there, I have a slightly simpler way of coming up with the g, h, j, k coefficients if you're interested in adopting that here. I confirmed that what you have is equivalent to what's there (thanks to sympy).

eepeterson replied:
Ok I just copied the cylinder from points method into this section, thanks!

paulromano commented on openmc/surface.py:
```diff
@@ -832,36 +778,79 @@ def z0(self, z0):
+        else:
+            a, b, c, d, e, f, g, h, j, k = coeffs
+
+        A = np.asarray([[a, d/2, f/2], [d/2, b, e/2], [f/2, e/2, c]])
```
If you know the object you're passing is _not_ an array, you might as well go directly to `np.array`
```suggestion
        A = np.array([[a, d/2, f/2], [d/2, b, e/2], [f/2, e/2, c]])
```

paulromano commented on openmc/surface.py:
```diff
@@ -832,36 +778,79 @@ def z0(self, z0):
+            a, b, c, d, e, f, g, h, j, k = coeffs
+
+        A = np.asarray([[a, d/2, f/2], [d/2, b, e/2], [f/2, e/2, c]])
+        bvec = np.asarray([g, h, j])
```
```suggestion
        bvec = np.array([g, h, j])
```

paulromano commented on openmc/surface.py:
```diff
@@ -1662,31 +1682,42 @@ def r2(self, r2):
+        g = -2*(dx*dx*x0 + dx*dy*y0 + dx*dz*z0 - c1)
+        h = -2*(dy*dy*y0 + dx*dy*x0 + dy*dz*z0 - c1)
+        j = -2*(dz*dz*y0 + dx*dz*x0 + dy*dz*y0 - c1)
+        k = (dx*x0 + dy*y0 + dz*z0)**2 - c1*(x0*x0 + y0*y0 + z0*z0)
```
Likewise, this block warrants a comment about how one comes up with the coefficients

eepeterson commented:
@paulromano, thanks for the review! I think I got all the changes you requested. Also for the general Cone and Cylinder, I implemented essentially a cache for the Quadric coefficients because it's more efficient to evaluate those geometries that way and would add a lot of flops if we had to compute them over and over. I think I'll do something similar for them in the C++ layer as well where it will make more of a difference.

paulromano commented:
Thanks @eepeterson. I'm not sure I understand the motivation for caching on the Python side. Why would the `_get_base_coeffs` method be called repeatedly? In normal use, it seems that it would only be called if a user wanted to translate a surface.

paulromano commented on openmc/surface.py:
```diff
@@ -268,62 +332,108 @@ def from_hdf5(group):
+        self._periodic_surface = periodic_surface
+        periodic_surface._periodic_surface = self
+
+    def _get_base_coeffs(self):
```
Should this one just be removed altogether?

eepeterson replied:
It's possibly starting to look that way and I'm all for simplifying wherever we can. Since almost every class has `x0, y0, z0` or something close to that translations should be able to be applied very easily without need for abstraction. I think we could even do the rotations for almost all the classes without abstraction since, but it might come back to bite us if we get rid of it. I need to think on this some more...

paulromano replied:
My point with this one is that `_get_base_coeffs` now appears as an `@abstractmethod` on `Surface`, from which all concrete surface classes derive. This method here in `PlaneMixin` is empty and doesn't functionally do anything. The same thing goes for `QuadricMixin`.

eepeterson replied:
Ah yes, I see what you're saying, I totally misinterpreted your comment.

eepeterson commented:
It also gets called in the evaluate method for general cones and cylinders which I was trying to make identical to how it should be implemented in the C++ layer

paulromano commented:
We also don't use that `evaluate` method on the Python side much, so I would suggest that in the interest of simplicity, let's leave out the caching for now.

eepeterson commented:
sounds good.

paulromano commented:
Thanks @eepeterson for this non-trivial refactoring!
